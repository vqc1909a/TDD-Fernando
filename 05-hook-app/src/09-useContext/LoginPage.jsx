import { useContext } from "react"
import { UserContext } from "./context/UserContext";

export const LoginPage = () => {

  const {user, setUser} = useContext(UserContext);
  return (
    <>
        <h1>LoginPage</h1>
        <hr />
        <pre aria-label="pre_login">
          { JSON.stringify(user, null, 3)}
        </pre>

        <button aria-label="button_login" className="btn btn-primary" onClick={() => setUser({id: 123, name: 'Juan', email: 'juan@google.com'})}>Set user</button>
    </>   
  )
}
