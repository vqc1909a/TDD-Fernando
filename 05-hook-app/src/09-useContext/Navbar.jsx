import { NavLink } from "react-router-dom"

export const Navbar = () => {
  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
        <div className="container-fluid">
            <NavLink className="navbar-brand" to="/">USE CONTEXT</NavLink>

            <div className="collapse navbar-collapse" id="navbarNav">
                <ul className="navbar-nav">
                    <li className="nav-item">
                        <NavLink className={(args) => {
                            // console.log(args)
                            // { isActive: true, isPending: false, isTransitioning: false }
                            return `nav-link ${args.isActive ? 'active' : ''}`
                        }} to="/">Home</NavLink>
                    </li>

                    <li className="nav-item">
                        <NavLink className={(args) => {
                            // console.log(args)
                            return `nav-link ${args.isActive ? 'active' : ''}`
                        }} to="/about">About</NavLink>
                    </li>
                    <li className="nav-item">
                        {/* We use the componente NavLink instead use Link because we can add the class "active" when the link has matched with the path url  */}
                        <NavLink className={(args) => {
                            // console.log(args)
                            return `nav-link ${args.isActive ? 'active' : ''}`
                        }
                        } to="/login">Login</NavLink>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
)
}
