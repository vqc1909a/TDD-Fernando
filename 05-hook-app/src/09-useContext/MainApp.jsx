import { Navigate, Route, Routes } from "react-router-dom";
import { AboutPage, HomePage, LoginPage } from "./";
import { Navbar } from "./Navbar";
import { UserProvider } from "./context/UserProvider";



export const MainApp = () => {
  return (
    // This UserProvider will should be in the rootComponent called main.jsx, it is here only for the example
    <UserProvider>

        <h1>MainApp</h1>
        <Navbar />
        <hr />

        <Routes>
          <Route path="/" element={<HomePage />} />
          <Route path="about" element={<AboutPage />} />
          <Route path="login" element={<LoginPage />} />
          <Route path="/*" element={<Navigate to="/about" />}/>
        </Routes>
    </UserProvider>
  )
}
