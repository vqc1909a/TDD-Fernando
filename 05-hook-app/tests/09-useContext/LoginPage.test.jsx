import { LoginPage } from "../../src/09-useContext/LoginPage"
import { fireEvent, render, screen } from "@testing-library/react"
import { UserContext } from "../../src/09-useContext/context/UserContext"
describe('tests on LoginPage component', () => { 

    const setUser = jest.fn();
    beforeEach(() => {
        jest.clearAllMocks();
    })
    test('should show the component by default', () => { 
        const { container } = render(
        <UserContext.Provider value={{user: undefined, setUser}}>
            <LoginPage />
        </UserContext.Provider>
        )

        const preTag = screen.getByLabelText('pre_login');
        expect(preTag.textContent).toBeFalsy();
        expect(preTag.textContent).toBe("");
        expect(container).toMatchSnapshot();
    })

    test('should call the function setUser when the trigger button click', () => {
        render(
        <UserContext.Provider value={{user: undefined, setUser}}>
            <LoginPage />
        </UserContext.Provider>
        )
        const buttonTag = screen.getByLabelText('button_login');
        fireEvent.click(buttonTag);
        expect(setUser).toHaveBeenCalled();
        expect(setUser).toHaveBeenCalledTimes(1);
        expect(setUser).toHaveBeenCalledWith({id: 123, name: 'Juan', email: 'juan@google.com'});
    })
})