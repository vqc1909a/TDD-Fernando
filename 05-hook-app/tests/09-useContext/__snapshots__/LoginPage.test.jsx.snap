// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on LoginPage component should show the component by default 1`] = `
<div>
  <h1>
    LoginPage
  </h1>
  <hr />
  <pre
    aria-label="pre_login"
  />
  <button
    aria-label="button_login"
    class="btn btn-primary"
  >
    Set user
  </button>
</div>
`;
